src/constants/util.rs

```diff
@@ -1,37 +1,28 @@
 use ff::PrimeField;
 use halo2::arithmetic::{CurveAffine, FieldExt};
 
-/// Decompose a scalar into FIXED_BASE_WINDOW_SIZE bits (little-endian)
+/// Decompose a scalar into `window_num_bits` bits (little-endian)
 /// For a window size of `w`, this returns [k_0, ..., k_n] where each `k_i`
 /// is a `w`-bit value, and `scalar = k_0 + k_1 * w + k_n * w^n`.
+/// Note that we are returning a `Vec<u8>` which means the window size is
+/// limited to <= 8 bits.
 pub fn decompose_scalar_fixed<C: CurveAffine>(
     scalar: C::Scalar,
     scalar_num_bits: usize,
     window_num_bits: usize,
 ) -> Vec<u8> {
-    let mut bits: Vec<bool> = scalar
+    // Pad bits to multiple of window_num_bits
+    let padding = (window_num_bits - (scalar_num_bits % window_num_bits)) % window_num_bits;
+    let bits: Vec<bool> = scalar
         .to_le_bits()
         .into_iter()
         .take(scalar_num_bits)
+        .chain(std::iter::repeat(false).take(padding))
         .collect();
-
-    assert_eq!(bits.len(), scalar_num_bits);
-
-    // Pad bits to multiple of window_num_bits
-    bits.append(&mut vec![
-        false;
-        (window_num_bits
-            - (scalar_num_bits % window_num_bits))
-            % window_num_bits
-    ]);
+    assert_eq!(bits.len(), scalar_num_bits + padding);
 
     bits.chunks_exact(window_num_bits)
-        .map(|chunk| {
-            let mut chunk = chunk.iter();
-            *(chunk.next().unwrap()) as u8
-                + ((*(chunk.next().unwrap()) as u8) << 1)
-                + ((*(chunk.next().unwrap()) as u8) << 2)
-        })
+        .map(|chunk| chunk.iter().rev().fold(0, |acc, b| (acc << 1) + (*b as u8)))
```

Comment:
I tried reversing the last two commits locally (to run the proptest on this buggy code) and verified that it indeed found this problem (and that this change fixes it).